/var/snap/nextcloud/current/mysql/mysql.sock exists after reboot

I was having some trouble in recovering nextcloud after a softlockup on a node. The nextcloud snap was running in a VM at the time of the hardware node hang and the hard reboot that followed.

So I got this in the journal, which is kind of confusing as there are multiple errors (`--initialize` seems ok as it is in the `start_mysql` wrapper):
```
дек 19 20:06:17 myvmhostname systemd[1]: Started Service for snap application nextcloud.mysql.
дек 19 20:06:17 myvmhostname snap[3892]: 2016-12-19T17:06:17.270288Z 0 [Warning] Changed limits: max_open_files: 1024 (requested 5000)
дек 19 20:06:17 myvmhostname snap[3892]: 2016-12-19T17:06:17.270329Z 0 [Warning] Changed limits: table_open_cache: 431 (requested 2000)
дек 19 20:06:17 myvmhostname snap[3892]: 2016-12-19T17:06:17.270426Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server option (see documentation for more details).
дек 19 20:06:17 myvmhostname snap[3892]: 2016-12-19T17:06:17.271004Z 0 [ERROR] --initialize specified but the data directory has files in it. Aborting.
дек 19 20:06:17 myvmhostname snap[3892]: 2016-12-19T17:06:17.271389Z 0 [ERROR] Aborting
дек 19 20:06:17 myvmhostname snap[3892]: Starting MySQL
дек 19 20:06:17 myvmhostname snap[3892]: .
дек 19 20:06:17 myvmhostname snap[3892]: Waiting for server...
дек 19 20:06:17 myvmhostname snap[3892]: Checking/upgrading mysql tables if necessary...
дек 19 20:06:17 myvmhostname snap[3892]: mysql_upgrade: Got error: 2002: Can't connect to local MySQL server through socket '/var/snap/nextcloud/current/mysql/mysql.sock' (111) while connecting to the MySQL server
дек 19 20:06:17 myvmhostname snap[3892]: Upgrade process encountered error and will not continue.
дек 19 20:06:17 myvmhostname snap[3913]: MySQL server process #3827 is not running! ... failed!
```
The same output remained after doing `systemctl restart snap.nextcloud.mysql.service`.

I poked around this and found that a unix socket that mysqld listens on was present even though the instance was not.

```
stat /var/snap/nextcloud/current/mysql/mysql.sock
  File: '/var/snap/nextcloud/current/mysql/mysql.sock'
  Size: 0         	Blocks: 0          IO Block: 4096   socket
Device: fc00h/64512d	Inode: 98043970    Links: 1
Access: (0777/srwxrwxrwx)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2016-12-20 21:09:25.921717143 +0300
Modify: 2016-12-20 21:09:18.505794944 +0300
Change: 2016-12-20 21:09:18.505794944 +0300
 Birth: -
```

I then did

```rm /var/snap/nextcloud/current/mysql/mysql.sock```

And the unit started just fine:

```
systemctl status snap.nextcloud.mysql.service
● snap.nextcloud.mysql.service - Service for snap application nextcloud.mysql
   Loaded: loaded (/etc/systemd/system/snap.nextcloud.mysql.service; enabled; vendor preset: enabled)
   Active: active (running) since Вт 2016-12-20 21:23:51 MSK; 55s ago
  Process: 3629 ExecStop=/usr/bin/snap run --command=stop nextcloud.mysql (code=exited, status=0/SUCCESS)
 Main PID: 4114 (start_mysql)
   CGroup: /system.slice/snap.nextcloud.mysql.service
           ├─4114 /bin/sh /snap/nextcloud/383/bin/start_mysql
           ├─4129 /bin/sh /snap/nextcloud/383/bin/mysqld_safe --datadir=/var/snap/nextcloud/383/mysql --pid-file=/var/snap/nextcloud/383/mysql/myvmhostname.pid --lc-messages-dir=/snap/next
           ├─4300 /snap/nextcloud/383/bin/mysqld --basedir=/snap/nextcloud/383 --datadir=/var/snap/nextcloud/383/mysql --plugin-dir=/snap/nextcloud/383/lib/plugin --lc-messages-dir=/snap/n
           └─4392 sleep 1

дек 20 21:23:51 myvmhostname snap[4114]: 2016-12-20T18:23:51.245513Z 0 [Warning] Changed limits: table_open_cache: 431 (requested 2000)
дек 20 21:23:51 myvmhostname snap[4114]: 2016-12-20T18:23:51.245611Z 0 [Warning] TIMESTAMP with implicit DEFAULT value is deprecated. Please use --explicit_defaults_for_timestamp server op
дек 20 21:23:51 myvmhostname snap[4114]: 2016-12-20T18:23:51.246204Z 0 [ERROR] --initialize specified but the data directory has files in it. Aborting.
дек 20 21:23:51 myvmhostname snap[4114]: 2016-12-20T18:23:51.246707Z 0 [ERROR] Aborting
дек 20 21:23:51 myvmhostname snap[4114]: Starting MySQL
дек 20 21:23:51 myvmhostname snap[4114]: .
дек 20 21:23:51 myvmhostname snap[4114]: Waiting for server...
дек 20 21:23:52 myvmhostname snap[4114]: Checking/upgrading mysql tables if necessary...
дек 20 21:23:52 myvmhostname snap[4114]: Checking if update is needed.
дек 20 21:23:52 myvmhostname snap[4114]: This installation of MySQL is already upgraded to 5.7.11, use --force if you still need to run mysql_upgrade
```

Now the mysql instance is listening on the socket as it should:
```
ss -ax | grep -i mysql
u_str  LISTEN     0      80     /var/snap/nextcloud/383/mysql/mysql.sock 33295                 * 0
```

From what I can see after a little grepping is that the socket location is configured in root.ini. 

```
# grep -RiP 'mysql.*?sock' /var/snap/nextcloud/current
...
/var/snap/nextcloud/current/nextcloud/config/config.php:  'dbhost' => 'localhost:/var/snap/nextcloud/current/mysql/mysql.sock',
...
/var/snap/nextcloud/current/mysql/root.ini:socket=/var/snap/nextcloud/current/mysql/mysql.sock
```

https://github.com/nextcloud/nextcloud-snap/blob/f1ad91131b142704c7e16b038cdf58ba08e33004/src/mysql/start_mysql#L6

My impression is that the socket files should be located on **tmpfs** to not be persistent across reboots - so some alternative to storing at ```/var/run/mysqld/mysqld.sock``` like in a standalone installation should be made.

NOTE: on Ubuntu /var/run is a symlink to /run which is where a tmpfs is mounted.
readlink /var/run ; grep "`readlink /var/run`\s" /proc/mounts
/run
tmpfs /run tmpfs rw,nosuid,noexec,relatime,size=332592k,nr_inodes=415737,mode=755 0 0

The default standalone mysql configuration files have this:
/etc/mysql/mysql.conf.d/mysqld.cnf:socket		= /var/run/mysqld/mysqld.sock
/etc/mysql/debian.cnf:socket   = /var/run/mysqld/mysqld.sock

Therefore, storing socket files on a tmpfs is a proper way to handle these files.

I am going to think about how to fit tmpfs mounting into the current snap structure but any suggestions are welcome.

--- a/snapcraft.yaml
+++ b/snapcraft.yaml
@@ -122,6 +122,7 @@ parts:
     files:
       src/apache/scripts/*: bin/
       src/apache/conf/*: conf/
+      src/apache/utilities/*: utilities/
 
   nextcloud:
     plugin: dump
@@ -191,6 +192,7 @@ parts:
     files:
       src/redis/config/*: config/redis/
       src/redis/scripts/*: bin/
+      src/redis/utilities/*: utilities/
 
   # Copy over our PHP configuration file.
   php-customizations:
@@ -198,6 +200,7 @@ parts:
     files:
       src/php/config/*: config/php/
       src/php/scripts/*: bin/
+      src/php/utilities/*: utilities/
     stage-packages: [mawk]
 
   # Copy over our Nextcloud configuration files
@@ -206,6 +209,7 @@ parts:
     files:
       src/nextcloud/config/*: htdocs/config/
       src/nextcloud/scripts/*: bin/
+      src/nextcloud/utilities/*: utilities/
 
   # Download the boost headers for MySQL. Note that the version used may need to
   # be updated if the version of MySQL changes.
@@ -283,6 +287,7 @@ parts:
       src/mysql/start_mysql: bin/
       src/mysql/my.cnf: my.cnf
       src/mysql/mysql.server: support-files/
+      src/mysql/utilities/*: utilities/
 
   mdns-publisher:
     plugin: go

--- a/src/apache/conf/httpd.conf
+++ b/src/apache/conf/httpd.conf
@@ -33,7 +33,7 @@ Mutex pthread
 
 #
 # PidFile: Allows you to place the pidfile in a specific location.
-PidFile "${SNAP_DATA}/apache/httpd.pid"
+PidFile "${APACHE_PIDFILE}"
 
 #
 # Dynamic Shared Object (DSO) Support
@@ -155,7 +155,7 @@ Alias "/.well-known/acme-challenge" "${SNAP_DATA}/certs/certbot/.well-known/acme
 # Setup the proxy to PHP-FPM
 ProxyTimeout 3600
 <FilesMatch \.php$>
-    SetHandler "proxy:unix:${SNAP_DATA}/php/php-fpm.sock|fcgi://localhost/"
+    SetHandler "proxy:unix:${PHP_FPM_SOCKET}|fcgi://localhost/"
 </FilesMatch>
 
 <Proxy "fcgi://localhost/">

--- a/src/apache/scripts/httpd-wrapper
+++ b/src/apache/scripts/httpd-wrapper
@@ -1,6 +1,8 @@
 #!/bin/sh
 
 . $SNAP/utilities/https-utilities
+. $SNAP/utilities/apache-utilities
+. $SNAP/utilities/php-utilities
 
 params=""
 if certificates_are_active; then

--- a/src/apache/utilities/apache-utilities
+++ b/src/apache/utilities/apache-utilities
@@ -1,0 +1,24 @@
+#!/bin/sh
+
+export APACHE_PIDFILE="/tmp/pids/httpd.pid"
+
+mkdir -p -m 750 "$(dirname $APACHE_PIDFILE)"
+
+restart_apache_if_running()
+{
+	if [ -f "$APACHE_PIDFILE" ]; then
+		# Restart apache by stopping it and letting systemd start it again.
+		apache_pid=$(cat "$APACHE_PIDFILE")
+		echo -n "Restarting apache... "
+		output=$(httpd-wrapper -k stop 2>&1)
+		if [ $? -eq 0 ]; then
+			while kill -0 $apache_pid 2>/dev/null; do
+				sleep 1
+			done
+			echo "done"
+		else
+			echo "error"
+			echo "$output"
+		fi
+	fi
+}

--- a/src/https/utilities/https-utilities
+++ b/src/https/utilities/https-utilities
@@ -1,5 +1,7 @@
 #!/bin/sh
 
+. $SNAP/utilities/apache-utilities
+
 SNAP_CURRENT=$(dirname $SNAP_DATA)/current
 
 LIVE_CERTS_DIRECTORY=$SNAP_CURRENT/certs/live
@@ -20,8 +22,6 @@ CUSTOM_ENABLE_HSTS=$CUSTOM_DIRECTORY/hsts
 
 CERTBOT_DIRECTORY=$SNAP_CURRENT/certs/certbot
 CERTBOT_LIVE_DIRECTORY=$CERTBOT_DIRECTORY/config/live
-
-APACHE_PIDFILE=$SNAP_DATA/apache/httpd.pid
 
 # If this function is run multiple times it will replace the certificate
 # and key if they're already present.
@@ -133,23 +133,4 @@ run_certbot()
 	certbot --text --config-dir $CERTBOT_DIRECTORY/config \
 	               --work-dir $CERTBOT_DIRECTORY/work \
 	               --logs-dir $CERTBOT_DIRECTORY/logs $@
-} 
-
-restart_apache_if_running()
-{
-	if [ -f "$APACHE_PIDFILE" ]; then
-		# Restart apache by stopping it and letting systemd start it again.
-		apache_pid=$(cat "$APACHE_PIDFILE")
-		echo -n "Restarting apache... "
-		output=$(httpd-wrapper -k stop 2>&1)
-		if [ $? -eq 0 ]; then
-			while kill -0 $apache_pid 2>/dev/null; do
-				sleep 1
-			done
-			echo "done"
-		else
-			echo "error"
-			echo "$output"
-		fi
-	fi
 }

--- a/src/mysql/mysql.server
+++ b/src/mysql/mysql.server
@@ -43,6 +43,8 @@
 # If you change base dir, you must also change datadir. These may get
 # overwritten by settings in the MySQL configuration files.
 
+. $SNAP/utilities/mysql-utilities
+
 basedir=$SNAP
 datadir=$SNAP_DATA/mysql
 
@@ -60,7 +62,7 @@ lock_file_path="$lockdir/mysql"
 # The following variables are only set for letting mysql.server find things.
 
 # Set some defaults
-mysqld_pid_file_path=
+mysqld_pid_file_path="$MYSQL_PIDFILE"
 if test -z "$basedir"
 then
   basedir=/
@@ -280,7 +282,7 @@ case "$mode" in
     then
       # Give extra arguments to mysqld with the my.cnf file. This script
       # may be overwritten at next upgrade.
-      $bindir/mysqld_safe --datadir="$datadir" --pid-file="$mysqld_pid_file_path" --lc-messages-dir="$SNAP/share" --socket="$SNAP_DATA/mysql/mysql.sock" $other_args >/dev/null 2>&1 &
+      $bindir/mysqld_safe --datadir="$datadir" --pid-file="$mysqld_pid_file_path" --lc-messages-dir="$SNAP/share" --socket="$MYSQL_SOCKET" $other_args >/dev/null 2>&1 &
       wait_for_pid created "$!" "$mysqld_pid_file_path"; return_value=$?
 
       # Make lock for RedHat / SuSE

--- a/src/mysql/start_mysql
+++ b/src/mysql/start_mysql
@@ -1,9 +1,9 @@
 #!/bin/sh
 
+. $SNAP/utilities/mysql-utilities
+
 root_option_file="$SNAP_DATA/mysql/root.ini"
 nextcloud_password_file="$SNAP_DATA/mysql/nextcloud_password"
-mysqld_pid_file_path=$SNAP_DATA/mysql/`hostname`.pid
-mysql_socket_file_path="/var/snap/$SNAP_NAME/current/mysql/mysql.sock"
 new_install=false
 
 # Make sure the database is initialized (this is safe to run if already
@@ -32,7 +32,7 @@ if [ $new_install = true ]; then
 
 	# Save root user information
 	echo "[client]" >> $root_option_file
-	echo "socket=$mysql_socket_file_path" >> $root_option_file
+	echo "socket=$MYSQL_SOCKET" >> $root_option_file
 	echo "user=root" >> $root_option_file
 	chmod 600 $root_option_file
 
@@ -61,10 +61,7 @@ SQL
 fi
 
 # Wait here until mysql is running
-echo "Waiting for server..."
-while [ ! -f "$mysqld_pid_file_path" -o ! -S "$mysql_socket_file_path" ]; do
-	sleep 1
-done
+wait_for_mysql
 
 # Check and upgrade mysql tables if necessary. This will return 0 if the upgrade
 # succeeded, in which case we need to restart mysql.
@@ -74,10 +71,8 @@ if [ $? -eq 0 ]; then
 	echo "Restarting mysql server after upgrade..."
 	$SNAP/support-files/mysql.server restart
 
-	echo "Waiting for server to come back after upgrade..."
-	while [ ! -f "$mysqld_pid_file_path" -o ! -S "$mysql_socket_file_path" ]; do
-        	sleep 1
-	done
+	# Wait for server to come back after upgrade
+	wait_for_mysql
 fi
 
 # If this was a new installation, wait until the server is all up and running
@@ -89,7 +84,7 @@ fi
 
 # Wait here until mysql exits (turn a forking service into simple). This is
 # only needed for Ubuntu Core 15.04, as 16.04 supports forking services.
-mysqld_pid=$(cat "$mysqld_pid_file_path")
-while kill -0 $mysqld_pid 2>/dev/null; do
+pid=$(mysql_pid)
+while kill -0 "$pid" 2>/dev/null; do
 	sleep 1
 done

--- a/src/mysql/utilities/mysql-utilities
+++ b/src/mysql/utilities/mysql-utilities
@@ -1,0 +1,33 @@
+#!/bin/sh
+
+export MYSQL_PIDFILE="/tmp/pids/mysql.pid"
+export MYSQL_SOCKET="/tmp/sockets/mysql.sock"
+
+mkdir -p -m 750 "$(dirname $MYSQL_PIDFILE)"
+mkdir -p -m 750 "$(dirname $MYSQL_SOCKET)"
+
+mysql_is_running()
+{
+	[ -f "$MYSQL_PIDFILE" -a -S "$MYSQL_SOCKET" ]
+}
+
+wait_for_mysql()
+{
+	if ! mysql_is_running; then
+		echo -n "Waiting for MySQL... "
+		while ! mysql_is_running; do
+			sleep 1
+		done
+		echo "done"
+	fi
+}
+
+mysql_pid()
+{
+	if mysql_is_running; then
+		cat "$MYSQL_PIDFILE"
+	else
+		echo "Unable to get MySQL PID as it's not yet running" >&2
+		echo ""
+	fi
+}

--- a/src/nextcloud/config/autoconfig.php
+++ b/src/nextcloud/config/autoconfig.php
@@ -12,7 +12,7 @@ $AUTOCONFIG = array(
 
 'dbtype' => 'mysql',
 
-'dbhost' => 'localhost:'.$data_path.'/mysql/mysql.sock',
+'dbhost' => 'localhost:'.getenv('MYSQL_SOCKET'),
 
 'dbname' => 'nextcloud',
 

--- a/src/nextcloud/config/config.php
+++ b/src/nextcloud/config/config.php
@@ -47,7 +47,7 @@ $CONFIG = array(
 'memcache.locking' => '\OC\Memcache\Redis',
 'memcache.local' => '\OC\Memcache\Redis',
 'redis' => array(
-    'host' => '/var/snap/'.$snap_name.'/current/redis/redis.sock',
+    'host' => getenv('REDIS_SOCKET'),
     'port' => 0,
 ),
 );

--- a/src/nextcloud/scripts/nextcloud-cron
+++ b/src/nextcloud/scripts/nextcloud-cron
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-export NEXTCLOUD_CONFIG_DIR=$SNAP_DATA/nextcloud/config
+. $SNAP/utilities/nextcloud-utilities
 
 echo -n "Waiting for Nextcloud config dir... "
 while [ ! -d "$NEXTCLOUD_CONFIG_DIR" ]; do

--- a/src/nextcloud/scripts/occ
+++ b/src/nextcloud/scripts/occ
@@ -1,5 +1,14 @@
 #!/bin/sh
 
-export NEXTCLOUD_CONFIG_DIR=$SNAP_DATA/nextcloud/config
+. $SNAP/utilities/php-utilities
+. $SNAP/utilities/nextcloud-utilities
+
+if [ $(id -u) -ne 0 ]; then
+	echo "This utility needs to run as root"
+	exit 1
+fi
+
+# occ can't do much before PHP FPM is up and running
+wait_for_php
 
 php -c $SNAP/config/php $SNAP/htdocs/occ $*

--- a/src/nextcloud/scripts/setup_nextcloud
+++ b/src/nextcloud/scripts/setup_nextcloud
@@ -1,5 +1,9 @@
 #!/bin/sh
 
+. $SNAP/utilities/php-utilities
+. $SNAP/utilities/nextcloud-utilities
+. $SNAP/utilities/redis-utilities
+
 # Make sure nextcloud directory exists
 mkdir -p -m 750 $SNAP_COMMON/nextcloud
 
@@ -9,16 +13,11 @@ mkdir -p -m 750 $SNAP_COMMON/nextcloud/tmp
 # Make sure nextcloud extra-apps directory exists (for user apps)
 mkdir -p -m 750 $SNAP_DATA/nextcloud/extra-apps
 
-# Wait for PHP FPM to be up and running before continuing, since we need to make
-# sure we can use occ below.
-php_pid_file_path=$SNAP_DATA/php/php-fpm.pid
-echo "Waiting for PHP..."
-while [ ! -f "$php_pid_file_path" ]; do
-	sleep 1
-done
+# We need both PHP and redis up and running before we can use occ
+wait_for_php
+wait_for_redis
 
 # If this is a new install, make sure it's configured correctly
-NEXTCLOUD_CONFIG_DIR=$SNAP_DATA/nextcloud/config
 if [ ! -d "$NEXTCLOUD_CONFIG_DIR" ]; then
 	echo "Configuring nextcloud..."
 	cp -r $SNAP/htdocs/config $NEXTCLOUD_CONFIG_DIR
@@ -26,7 +25,7 @@ else
 	# This is not a new installation, so we don't want to overwrite the config.
 	# We do, however, want to make sure we incorporate the new capabilities of
 	# this snap version, namely, using Redis for the memcache and file locking.
-	occ config:system:set redis host --value="/var/snap/$SNAP_NAME/current/redis/redis.sock" --type=string
+	occ config:system:set redis host --value="$REDIS_SOCKET" --type=string
 	occ config:system:set redis port --value=0 --type=integer
 	occ config:system:set memcache.locking --value="\OC\Memcache\Redis" --type=string
 	occ config:system:set memcache.local --value="\OC\Memcache\Redis" --type=string

--- a/src/nextcloud/utilities/nextcloud-utilities
+++ b/src/nextcloud/utilities/nextcloud-utilities
@@ -1,0 +1,3 @@
+#!/bin/sh
+
+export NEXTCLOUD_CONFIG_DIR=$SNAP_DATA/nextcloud/config

--- a/src/php/config/php-fpm.conf
+++ b/src/php/config/php-fpm.conf
@@ -14,7 +14,7 @@
 ; Pid file
 ; Note: the default prefix is /home/ubuntu/src/nextcloud-snap/parts/php/install/var
 ; Default Value: none
-pid = ${SNAP_DATA}/php/php-fpm.pid
+pid = ${PHP_FPM_PIDFILE}
 
 ; Error log file
 ; If it's set to "syslog", log is sent to syslogd instead of being written

--- a/src/php/config/php-fpm.d/www.conf
+++ b/src/php/config/php-fpm.d/www.conf
@@ -33,7 +33,7 @@ group = root
 ;                            (IPv6 and IPv4-mapped) on a specific port;
 ;   '/path/to/unix/socket' - to listen on a unix socket.
 ; Note: This value is mandatory.
-listen = ${SNAP_DATA}/php/php-fpm.sock
+listen = ${PHP_FPM_SOCKET}
 
 ; Set listen(2) backlog.
 ; Default Value: 511 (-1 on FreeBSD and OpenBSD)

--- a/src/php/scripts/start-php-fpm
+++ b/src/php/scripts/start-php-fpm
@@ -1,17 +1,16 @@
 #!/bin/sh
 
+. $SNAP/utilities/mysql-utilities
+. $SNAP/utilities/php-utilities
+. $SNAP/utilities/redis-utilities
+
 mkdir -p -m 750 ${SNAP_DATA}/php
 
-mysqld_pid_file_path=$SNAP_DATA/mysql/`hostname`.pid
-# Wait for mysql to be up and running, since we need to make sure
-# we run the upgrade process.
-echo "Waiting for mysql..."
-while [ ! -f "$mysqld_pid_file_path" ]; do
-	sleep 1
-done
+# We need to make sure mysql is running so we can run the migration process
+wait_for_mysql
 
 # Wait until we have an nextcloud mysql password
-echo "Obtaining nextcloud mysql credentials..."
+echo -n "Obtaining nextcloud mysql credentials... "
 nextcloud_password_path=$SNAP_DATA/mysql/nextcloud_password
 timeout=10
 while [ $timeout -gt 0 -a ! -e $nextcloud_password_path ]; do
@@ -19,11 +18,13 @@ while [ $timeout -gt 0 -a ! -e $nextcloud_password_path ]; do
 	sleep 1
 done
 if [ ! -e $nextcloud_password_path ]; then
+	echo ""
 	echo -n "Timed out while attempting to obtain nextcloud mysql password. "
 	echo -n "This isn't unusual when starting up for the first time after "
 	echo "an install or an upgrade. Will try again."
 	exit 1
 fi
+echo "done"
 
 # Tends to be between 30-50MB
 average_php_memory_requirement=50

--- a/src/php/utilities/php-utilities
+++ b/src/php/utilities/php-utilities
@@ -1,0 +1,33 @@
+#!/bin/sh
+
+export PHP_FPM_PIDFILE="/tmp/pids/php-fpm.pid"
+export PHP_FPM_SOCKET="/tmp/sockets/php-fpm.sock"
+
+mkdir -p -m 750 "$(dirname $PHP_FPM_PIDFILE)"
+mkdir -p -m 750 "$(dirname $PHP_FPM_SOCKET)"
+
+php_is_running()
+{
+	[ -f "$PHP_FPM_PIDFILE" -a -S "$PHP_FPM_SOCKET" ]
+}
+
+wait_for_php()
+{
+	if ! php_is_running; then
+		echo -n "Waiting for PHP... "
+		while ! php_is_running; do
+			sleep 1
+		done
+		echo "done"
+	fi
+}
+
+php_pid()
+{
+	if php_is_running; then
+		cat "$PHP_FPM_PIDFILE"
+	else
+		echo "Unable to get PHP PID as it's not yet running" >&2
+		echo ""
+	fi
+}

--- a/src/redis/config/redis.conf
+++ b/src/redis/config/redis.conf
@@ -98,7 +98,7 @@ tcp-backlog 511
 # incoming connections. There is no default, so Redis will not listen
 # on a unix socket when not specified.
 #
-unixsocket ${SNAP_DATA}/redis/redis.sock
+unixsocket ${REDIS_SOCKET}
 # unixsocketperm 700
 
 # Close the connection after a client is idle for N seconds (0 to disable)
@@ -147,7 +147,7 @@ supervised no
 #
 # Creating a pid file is best effort: if Redis is not able to create it
 # nothing bad happens, the server will start and run normally.
-pidfile ${SNAP_DATA}/redis/redis.pid
+pidfile ${REDIS_PIDFILE}
 
 # Specify the server verbosity level.
 # This can be one of:

--- a/src/redis/scripts/start-redis-server
+++ b/src/redis/scripts/start-redis-server
@@ -1,8 +1,10 @@
 #!/bin/sh
 
+. $SNAP/utilities/redis-utilities
+
 mkdir -p -m 750 ${SNAP_DATA}/redis
 
 # redis doesn't support environment variables in its config files. Thankfully
 # it supports reading the config file from stdin though, so we'll rewrite the
 # config file on the fly and pipe it in.
-sed -e "s|\${SNAP_DATA}|$SNAP_DATA|" $SNAP/config/redis/redis.conf | redis-server -
+sed -e "s|\${SNAP_DATA}|$SNAP_DATA|;s|\${REDIS_PIDFILE}|$REDIS_PIDFILE|;s|\${REDIS_SOCKET}|$REDIS_SOCKET|" $SNAP/config/redis/redis.conf | redis-server -

--- a/src/redis/utilities/redis-utilities
+++ b/src/redis/utilities/redis-utilities
@@ -1,0 +1,33 @@
+#!/bin/sh
+
+export REDIS_PIDFILE="/tmp/pids/redis.pid"
+export REDIS_SOCKET="/tmp/sockets/redis.sock"
+
+mkdir -p -m 750 "$(dirname $REDIS_PIDFILE)"
+mkdir -p -m 750 "$(dirname $REDIS_SOCKET)"
+
+redis_is_running()
+{
+	[ -f "$REDIS_PIDFILE" -a -S "$REDIS_SOCKET" ]
+}
+
+wait_for_redis()
+{
+	if ! redis_is_running; then
+		echo -n "Waiting for redis... "
+		while ! redis_is_running; do
+			sleep 1
+		done
+		echo "done"
+	fi
+}
+
+redis_pid()
+{
+	if redis_is_running; then
+		cat "$REDIS_PIDFILE"
+	else
+		echo "Unable to get redis PID as it's not yet running" >&2
+		echo ""
+	fi
+}

